### PR TITLE
Add ability to serve static content/resources from any given directory

### PR DESCRIPTION
**Pull request by kipchubett**
Spring framework comes with a pre-configured implementation of [ResourceHttpRequestHandler](https://github.com/spring-projects/spring-framework/blob/master/spring-webmvc/src/main/java/org/springframework/web/servlet/resource/ResourceHttpRequestHandler.java) to facilitate serving static resources.

By default, this handler serves static content from any of /static, /public, /resources, and /META-INF/resources directories that are on the classpath. So, serving from a custom directory can easily be achieved by overriding the default handlers configuration i.e

```java
@Override
public void addResourceHandlers(ResourceHandlerRegistry registry) {
    registry
      .addResourceHandler("/files/**")
      .addResourceLocations("file:/opt/files/");
 }
```
However, that doesn't work for reasons that I am not sure 🤷🏽‍♂️ . `ResourceHttpRequestHandler` was introduced in [spring 3.0.4](https://github.com/spring-projects/spring-framework/blob/master/spring-webmvc/src/main/java/org/springframework/web/servlet/resource/ResourceHttpRequestHandler.java#L97).

So I use [ResourceLoaders](https://docs.spring.io/spring-framework/docs/5.2.4.RELEASE/spring-framework-reference/core.html#resources-resourceloader) instead, worked pretty well. With ResourceLoader, you can serve static content/resources from;
   - Classpath - `frontend/`
   - Filesystem - `file:c:/opt/openmrs-frontend/`
   - Directly from any URL - `https://openmrs.org/fronted/`
   - Main application directory

This is the key configuration 👇🏾
Screenshot 2020-12-07 at 10 05 53

### CODE REVIEW

**Review by FlorianRappl — APPROVED** [review]
LGTM - thanks a lot @corneliouzbett 🍻 !

**Comment by dkayiwa** [comment]
Was the existing implementation hard coding a specific directory?

**Comment by FlorianRappl** [comment]
> Was the existing implementation hard coding a specific directory?

I think yes. It was the webapp folder (https://github.com/openmrs/openmrs-module-spa/tree/master/omod/src/main/webapp).

File: omod/src/main/java/org/openmrs/module/spa/servlet/SpaServlet.java [hunk 1]
```diff
@@ -1,51 +1,101 @@
...
+     *
+     * @see HttpServlet#getLastModified(HttpServletRequest)
+     */
+    @Override
+    protected long getLastModified(HttpServletRequest req) {
+        File f = null;
+        try {
+            f = getResource(req).getFile();
+        } catch (IOException e) {
+            e.printStackTrace();
```
Comment (dkayiwa) [inline omod/src/main/java/org/openmrs/module/spa/servlet/SpaServlet.java]: https://wiki.openmrs.org/display/docs/Java+Conventions#JavaConventions-Donotprintout

File: omod/src/main/java/org/openmrs/module/spa/utils/SpaModuleUtils.java [hunk 2]
```diff
@@ -20,6 +33,7 @@
      * to valid file or directory
      * @return directory for the path specified in the GP
      */
+    @Deprecated
```
Comment (dkayiwa) [inline omod/src/main/java/org/openmrs/module/spa/utils/SpaModuleUtils.java]: https://wiki.openmrs.org/display/docs/Java+Conventions#JavaConventions-Deprecation

File: omod/src/main/java/org/openmrs/module/spa/servlet/SpaServlet.java [hunk 3]
```diff
@@ -1,51 +1,101 @@
...
+        }
+    }
+
+    /**
+     * Turns the given request/path into a File object
+     *
+     * @param request the current http request
+     * @return the file being requested or null if not found
+     */
+    protected Resource getResource(HttpServletRequest request) {
```
Comment (dkayiwa) [inline omod/src/main/java/org/openmrs/module/spa/servlet/SpaServlet.java:82]: Would be great to add some tests for this method.

**Comment by kipchubett** [comment]
@dkayiwa PTAL

**Comment by dkayiwa** [comment]
Thanks @corneliouzbett

**Comment by dkayiwa** [comment]
@corneliouzbett do you want to set the scope to `provided` for `lombok`?

**Comment by brandones** [comment]
> So I use ResourceLoaders instead, worked pretty well. With ResourceLoader, you can serve static content/resources from;
>
>  Classpath - frontend/
    Filesystem - file:c:/opt/openmrs-frontend/
    Directly from any URL - https://openmrs.org/fronted/
    Main application directory

@corneliouzbett How can I serve from the "Main application directory"? That's the original use case this module was built for.

**Pull request merged**